### PR TITLE
Copter: Use explicit dependencies instead of hardcoded numbers

### PR DESCRIPTION
**Pull request by amilcarlucas**
This is a remain of #7458

### CODE REVIEW

File: ArduCopter/Copter.h [hunk 1]
```diff
@@ -211,6 +211,7 @@ class Copter : public AP_HAL::HAL::Callbacks {
     ParametersG2 g2;
 
     // main loop scheduler
+#define MAIN_LOOP_SCHEDULER_HZ 400
```
Comment (peterbarker) [inline ArduCopter/Copter.h]: Soooo.... what happens when someone sets `SCHED_LOOP_RATE` to something else?  It's really going to be better to just stash away a `_roi_yaw_last_updated_ms` value here, I think.

Also, I'm developing a pet hate of `#define`s for numeric constants :-)    @WickedShell is promoting having these values directly in the code where they are only used once - so in this case you'd have a `const uint8_t _yaw_update_interval_ms = 10;` in the `roi_yaw` method.

Comment (muramura) [reply ArduCopter/Copter.h]: I think that it is better to refer to this value as the loop rate value.
libraries\AP_Scheduler\AP_Scheduler.cpp
    // @Param: LOOP_RATE
    // @DisplayName: Scheduling main loop rate
    // @Description: This controls the rate of the main control loop in Hz. This should only be changed by developers. This only takes effect on restart. Values over 400 are considered highly experimental.
    // @Values: 50:50Hz,100:100Hz,200:200Hz,250:250Hz,300:300Hz,400:400Hz
    // @RebootRequired: True
    // @User: Advanced
    AP_GROUPINFO("LOOP_RATE",  1, AP_Scheduler, _loop_rate_hz, SCHEDULER_DEFAULT_LOOP_RATE),

**Review by khancyr — CHANGES_REQUESTED** [review]
as per comment on https://github.com/ArduPilot/ardupilot/issues/11187, it was asked to remove the counter and test back.

**Comment by amilcarlucas** [comment]
I think I have addressed the comments now

File: ArduCopter/Copter.cpp [hunk 2]
```diff
@@ -513,6 +513,9 @@ void Copter::one_hz_loop()
     // log terrain data
     terrain_logging();
 
+    // Update the ROI yaw loop rate divider if the user changes the mail loop rate
```
Comment (khancyr) [inline ArduCopter/Copter.cpp:516]: *main ?

File: ArduCopter/autoyaw.cpp [hunk 3]
```diff
@@ -2,11 +2,19 @@
 
 Mode::AutoYaw Mode::auto_yaw;
 
+#define ROI_YAW_SCHEDULER_HZ 100 // ROI yaw should be calculated at 100Hz
```
Comment (khancyr) [inline ArduCopter/autoyaw.cpp:5]: not blocking but I prefer static constexpr uint8_t here,

**Review by khancyr — APPROVED** [review]
Looks good!

**Comment by peterbarker** [comment]
@amilcarlucas just came across this one again.  We've now declared that the main loop rate is a reboot-required parameter (good!), so I'm sorry my comment here appears to have moved this away from merging rather than towards it.

Your new variable here would be set once at boot, I think.

@rmackay9 would you accept a patch which allows the ROI yaw update to work the same regardless of main loop rate - what this is trying to do?

File: ArduCopter/Copter.cpp (hunk 2 above)
Comment (amilcarlucas) [reply ArduCopter/Copter.cpp:516]: good point

Comment (amilcarlucas) [reply ArduCopter/Copter.cpp:516]: done

File: ArduCopter/autoyaw.cpp (hunk 3 above)
Comment (amilcarlucas) [reply ArduCopter/autoyaw.cpp]: done

File: ArduCopter/Copter.h (hunk 1 above)
Comment (amilcarlucas) [reply ArduCopter/Copter.h]: done

**Comment by amilcarlucas** [comment]
@rmackay9 should we go this way, or should we run the yaw controller at full 400Hz ?

**Comment by rmackay9** [comment]
Let's run it at full rate.  We could perhaps measure the CPU cost but I don't think it will be too bad.

**Comment by amilcarlucas** [comment]
@rmackay9 I implemented your suggestion at #16523

**Comment by amilcarlucas** [comment]
Closing now that #16523 got merged

**Pull request closed without merging**